tools/autograd/templates/Functions.h

```diff
@@ -46,6 +46,31 @@ struct TypeAndSize {
   at::TensorOptions options;
 };
 
+class StridesOrErrorIfSparse {
```

Comment:
The basic purpose of this class is to either:
* Hold onto and return a tensor's property (in this case `strides`) when a backward method that uses that property is called.
* Throw an error if that property is not accessible (in this case, calling `strides()` on a sparse tensor throws an error).

This means that we are no longer referencing `mat1` for the `mm_mat1_backward` call because the forward method pulls out all the information it needs from `mat1`. Avoiding a reference to `mat1` is a requirement for the next step, which is to avoid keeping any reference to `mat1` in the backward kernel when `mat2` does not require grad.

I imagine that if we can avoid unnecessary references to `mat1` in `MmBackward`, there are probably other backward methods that also have avoidable references. So I wonder if we would need a more generalized class than `StridesOrErrorIfSparse`--a class that can return any property of a tensor or error on any condition.